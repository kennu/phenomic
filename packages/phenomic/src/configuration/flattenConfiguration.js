import path from "path"

const debug = require("debug")("phenomic:core:configuration")

const normalizeModule = (module) => {
  if (typeof module === "string") {
    module = require(path.resolve(path.join("node_modules", module)))
  }

  // for es6 transpiled code
  if (module && typeof module.default === "function") {
    module = module.default
  }

  return module
}

function flattenConfiguration(config: PhenomicInputPreset = {}): Array<(arg: ?any) => PhenomicPlugin> {
  const plugins = [
    ...(config.presets || [])
      .map(normalizeModule)
      .reduce((acc, preset) => {
        acc.push(...flattenConfiguration(preset()))
        return acc
      }, []),
    ...(config.plugins || []).map(normalizeModule),
  ]
  debug("flattenConfiguration", plugins.map((plugin) => plugin.name))
  return plugins
}

export default flattenConfiguration
