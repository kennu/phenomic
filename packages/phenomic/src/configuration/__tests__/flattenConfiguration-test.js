import flattenConfiguration from "../flattenConfiguration"

describe("flattenConfiguration", () => {

  it("should flatten correctly", () => {

    expect(
      flattenConfiguration({
        plugins: [ "a", "b" ],
      })
    ).toMatchSnapshot()

    expect(
      flattenConfiguration({
        plugins: [ "a", "b" ],
        presets: [
          {
            plugins: [ "c" ],
          },
        ],
      })
    ).toMatchSnapshot()

    expect(
      flattenConfiguration({
        plugins: [ "a", "b" ],
        presets: [
          {
            plugins: [ "c" ],
            presets: [
              {
                plugins: [ "d" ],
              },
            ],
          },
        ],
      })
    ).toMatchSnapshot()

  })

})
