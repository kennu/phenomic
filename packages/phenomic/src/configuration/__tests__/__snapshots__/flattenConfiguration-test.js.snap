exports[`flattenConfiguration should flatten correctly 1`] = `
Array [
  "a",
  "b",
]
`;

exports[`flattenConfiguration should flatten correctly 2`] = `
Array [
  "a",
  "b",
  "c",
]
`;

exports[`flattenConfiguration should flatten correctly 3`] = `
Array [
  "a",
  "b",
  "c",
  "d",
]
`;
